design/authenticators_health_api.md

```diff
@@ -0,0 +1,231 @@
+
+##### Enabled
+The "enabled" authenticators are those who are whitelisted in the CONJUR_AUTHENTICATORS variable. If the variable is not configured then we have only the "authn" authenticator.
+
+##### Conclusion
+It seems that we'll want to list all the "installed" authenticators and test if they are healthy.
+
+Note: At this point the "authn-oidc" is not showing in the installed authenticators as it doesn't meet 
+the first & third sections in the criteria. we'll need to fix this before implementing this.
+
+### Incremental health check for authenticators
+
+As each authenticator health check will need its own development effort, we'll want to have this feature in production even if we didn't perform the needed development for each authenticator. This means that we need some "N/A" status for authenticators which we didn't develop their health-check.
+
+So the statuses for authenticators are:
+
+- ok: the authenticator is healthy
+- fail: the authenticator is unhealthy
+- not-implemented: he authenticator health-check is not implemented
+
+So a response can look like this:
+```
+{
+   “authenticators”:
+   {
+      "authn-ldap": "not-implemented",
+      "authn-oidc": "fail",
+      "authn": "ok"
+   },
+   "ok": false
+}
+```
+
+Another option is to not have the authenticator health-check in the response at all in case it's not implemented. so for the case above, the response will be:
+```
+{
+   “authenticators”:
+   {
+      "authn-oidc": "fail",
+      "authn": "ok"
+   },
+   "ok": false
+}
+```
+
+I think the first option is better as it tells the full story explicitly without leaving the user with questions.
+
+### Response code for unhealthy authenticators
+
+In case of an error we need to return a response code which will best indicate the issue. Let's explore the possibilities and find the best response code. 
+
+A valid option will be to return an error response code (4xx, 5xx). In this case the 500's make more sense as the error here is not in the actual request so it's more of a server error that a client error. The possibilities for 5xx responses are:
+
+- 500 Internal Server Error
+    - Con: This is very general and we should try to avoid it when possible
+- 501 Not Implemented
+    - Con: Well, this is actually implemented. 
+- 503 Service Unavailable
+    - Con: This doesn't really tell the story as the service that we're calling - The Authenticator health service - is actually available. 
+- 502 Bad Gateway & 504 Gateway Timeout
+    - Irrelevant as this is not a gateway.
+
+So response codes from the 4xx & the 5xx families aren't quite good.
+
+Another option is to return a 200 OK response. Yes, there is an error and 200 doesn't indicate that but the response body will. This also makes sense as the request itself succeeded and the server returned a response successfully so it makes sense to return a successful code.
+
+Conclusion: return 200 OK. 
```

Comment:
200 should never be returned for faulty components. 500 is general but you can add messages to the response that will elaborate on what the issue is in both html and json formats.

Reply:
i get your point about 200. i'll move to 5xx.

i noticed now that the general health-check returns `502 Bad Gateway` in case one of the services is unavailable. While it makes sense there in some way as conjur is being used as a gateway to the services, here it seems pretty drunk so i have to agree with you about 500. I'll change it in the doc.